test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/RequestTests.cs

```diff
@@ -76,6 +71,65 @@ public class RequestTests
             }
         }
 
+        [ConditionalFact]
+        [FrameworkSkipCondition(RuntimeFrameworks.Mono, SkipReason = "Needs to run on 64-bit Mono, which isn't available on OSX.")]
+        public async Task LargeMultipartUpload()
+        {
+            var port = PortManager.GetPort();
+            var builder = new WebHostBuilder()
+                .UseServer("Microsoft.AspNetCore.Server.Kestrel")
+                .UseUrls($"http://localhost:{port}/")
+                .Configure(app =>
+                {
+                    app.Run(async context =>
+                    {
+                        long total = 0;
+                        var bytes = new byte[1024];
+                        var count = await context.Request.Body.ReadAsync(bytes, 0, bytes.Length);
+                        while (count > 0)
+                        {
+                            total += count;
+                            count = await context.Request.Body.ReadAsync(bytes, 0, bytes.Length);
+                        }
+                        await context.Response.WriteAsync(total.ToString(CultureInfo.InvariantCulture));
+                    });
+                });
+
+            using (var host = builder.Build())
+            {
+                host.Start();
+
+                using (var client = new HttpClient())
+                {
+                    using (var form = new MultipartFormDataContent())
+                    {
+                        const int oneGigabyte = 1024 * 1024 * 1024;
+                        const int files = 2;
+
+                        for (int i = 0; i < files; i++)
+                        {
+                            var fileName = Guid.NewGuid().ToString();
+                            var bytes = new byte[oneGigabyte];
+                            var content = new ByteArrayContent(bytes);
+                            content.Headers.ContentDisposition = new ContentDispositionHeaderValue("form-data")
+                            {
+                                Name = "\"files\"",
+                                FileName = "\"" + fileName + "\""
+                            };
+                            content.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
+
+                            form.Add(new ByteArrayContent(bytes), "file", fileName);
```

Comment:
We already use Socket's to hand roll requests in our `EngineTests`, so hopefully this won't be much more difficult. I think it's worth it to avoid such high memory usage.